onnx/test/cpp/parser_test.cc

```diff
@@ -67,6 +68,37 @@ TEST(ParserTest, TypeTest) {
   Parse(type, "float[N,?,K]");
   EXPECT_FALSE(type.tensor_type().shape().dim(1).has_dim_param());
   EXPECT_FALSE(type.tensor_type().shape().dim(1).has_dim_value());
+
+  // sequence type:
+  Parse(type, "seq(float[])");
+  EXPECT_TRUE(type.has_sequence_type());
+  auto& elttype = type.sequence_type().elem_type();
+  EXPECT_TRUE(elttype.has_tensor_type());
+  EXPECT_EQ(elttype.tensor_type().elem_type(), float_type);
+  EXPECT_FALSE(elttype.tensor_type().has_shape());
+
+  // optional type:
+  Parse(type, "optional(float)");
+  EXPECT_TRUE(type.has_optional_type());
+  auto& optelttype = type.optional_type().elem_type();
+  EXPECT_TRUE(optelttype.has_tensor_type());
+  EXPECT_EQ(optelttype.tensor_type().elem_type(), float_type);
+  EXPECT_TRUE(optelttype.tensor_type().has_shape());
+
+  // optional type:
+  Parse(type, "sparse_tensor(float[1000])");
```

Comment:
Maybe we can extend the test to cover all cases mentioned in
```
            // Grammar:
            // float indicates scalar (rank 0)
            // float [] indicates unknown rank tensor (not a zero rank tensor)
            // float [one-or-more-dimensions] indicates tensor of known rank > 0.
```
?

Reply:
Those are already tested separately. Since the parser is recursive, I think the combinations should work fine. Not sure we need to enumerate all possibilities. However, I did add different forms for the different constructors (like seq, map, optional) so overall there is good coverage, I think.